pandas/core/arrays/categorical.py

```diff
@@ -314,6 +314,16 @@ class Categorical(ExtensionArray, PandasObject):
     def __init__(self, values, categories=None, ordered=None, dtype=None,
                  fastpath=False):
 
+        # GH23814, for perf, if no optional params used and values already an
```

Comment:
i think we can just move this down to where the fastpath check is now; you can add this on i think. this constructor is already amazing *too* complicated.

Reply:
@eoveson I would still like to investiagte consolidating some of this code. This is a very complicated constructor and more code is not great here. See if you can add it lower down, even if its slightly lower perf.

Reply:
Updated, please check it out when you get a chance